File: selenide_testng_simple_with_wdm/src/test/java/pages/LoginPage.java
package pages;

import com.codeborne.selenide.SelenideElement;
import objects.User;
import org.openqa.selenium.By;

import static com.codeborne.selenide.Condition.*;
import static com.codeborne.selenide.Selenide.$;
import static com.codeborne.selenide.Selenide.open;

public class LoginPage {

    private SelenideElement usernameField = $(By.id("user-name"));
    private SelenideElement passwordField = $(By.id("password"));
    private SelenideElement loginButton = $(By.id("login-button"));
    private SelenideElement errorMsgContainer = $("h3[data-test='error']");

    public LoginPage() {
        open("/");
    }

    public void authenticateUser(User user) {
        usernameField.should(be(visible)).val(user.getUsername());
        passwordField.should(be(visible)).val(user.getPassword());
        loginButton.should(be(visible)).click();
    }

    public void verifyErrorMessage(String expectedMsg) {
        errorMsgContainer.shouldHave(text(expectedMsg));
    }

    public void errorIsNotDisplayed(){
        errorMsgContainer.should(not(be(visible)));
    }

}

File: cypress_12_with_gherkin/cypress/pages/login.page.js
/// <reference types="cypress" />

import test_credentials from '../fixtures/test_credentials.json'

export class LoginPage {

    locators = {
        username: () => cy.get('#user-name'),
        password: () => cy.get('#password'),
        loginBtn: () => cy.get('#login-button'),
        loginLogo: () => cy.get('.login_logo'),
        errorMsgContainer: () => cy.get('h3[data-test="error"]')
    }

    visitLoginPage(){
        cy.visit("/")
    }

    confirmLoginPageDisplayed(){
        this.locators.loginBtn().should('be.visible')
        this.locators.username().should('be.visible')
        this.locators.password().should('be.visible')
    }

    authenticateUser(user, pass){
        if(user === undefined){
            user = test_credentials.valid_username
        }
        if (pass === undefined) {
            pass = test_credentials.valid_password
        }
        this.locators.username().type(user)
        this.locators.password().type(pass)
        this.locators.loginBtn().click()
    }

    validateInvalidLoginMsg(msg){
        this.locators.errorMsgContainer().should('have.text', msg)
    }

}